src/Oscoin/Storage/Block/SQLite/Internal.hs

```diff
@@ -58,29 +59,54 @@ import           Codec.Serialise
 import           Oscoin.Storage.Block.Cache (BlockCache)
 import qualified Oscoin.Storage.Block.Cache as BlockCache
 
+-- Simple, composable transaction mechanism to avoid shooting ourselves in the
+-- foot when working with MVars. It ensures that 'withMVar' is called in a
+-- single place ('runTransaction') rather than being scattered in multiple places.
+-- It also allows transactions to be composed together.
+newtype Transaction m a = TransactionT (ReaderT Connection m a)
+    deriving (Functor, Applicative, Monad, MonadReader Connection, MonadIO)
+
 -- | A handle to an on-disk block store.
 data Handle c tx s = Handle
-    { hConn       :: Connection                    -- ^ Connection to on-disk storage for non-orphan blocks.
+    { hConn       :: MVar Connection               -- ^ Connection to on-disk storage for non-orphan blocks.
```

Comment:
How about `newtype`ing `MVar Connection`?

Reply:
Can do! What about `ConnectionLock` or something like that?

Reply:
I'm not sure ... does it need to "leave" this module? I think just `Connection` would be fine.

Reply:
At the moment yes, it does leave this module (cfr. the `WhiteBox` tests). I can simply type alias this though, if this will still aid the visibility/readability within the `Internal` module.